uses BufferWriter to compose format string

d2r opened the pull request:
Milestone: 8.0.0
Labels: Backports, Logging
Project: 7.x releases

Backport:
* This depends on #3098.
* Backport requested for potential 7.2 release, not 7.x branch (to avoid needing to backport BufferWriter as well)

d2r commented:
It appears there is a problem with `test_Buffer` and `test_Event`.
Need to investigate.

d2r commented:
I attempted to reproduce the failures using the `ci.trafficserver.apache.org/ats/centos:7` docker image, but I did not see it.

Reopening to see if the issues seen earlier can be seen again.

d2r commented:
[approve ci]

d2r commented:
Looks like the cores may be related to a `SIGABRT` from an assert thrown from BufferWriter:
```
Fatal: ../../lib/ts/BufferWriter.h:311: failed assertion `_capacity < N`
FAIL test_Buffer (exit status: 134)
```
I am not able to reproduce this, but I'll take a look.

d2r commented:
Blocked on #3098.

d2r commented:
[approve ci]

d2r commented:
@SolidWallOfCode @bryancall

SolidWallOfCode commented on lib/ts/Diags.cc:
```diff
@@ -226,43 +214,50 @@ void
-  // format_buf contains <prefix_string>: (<debug_tag>) <format_string> //
-  // and format_buf_w_ts has the same thing with a prepended timestamp. //
-  ////////////////////////////////////////////////////////////////////////
+  const size_t BUF_CAPACITY = 1024;
```
Is this used anywhere else? If not, might as well just put the `1024` explicitly in the `LocalBufferWriter` declaration.

d2r replied:
Yes, we can clean this up. Formerly there were two buffers, and now there is one.

SolidWallOfCode commented on lib/ts/Diags.cc:
```diff
@@ -226,43 +214,50 @@ void
-  // add the thread id
-  end_of_format += snprintf(end_of_format, sizeof(format_buf), "{0x%" PRIx64 "} ", (uint64_t)ink_thread_self());
+    format_writer.write('[');
+    format_writer.write(buffer + 4, std::min(strlen(buffer + 4), sizeof(timestamp_buf) - 4));
```
Why not just `strlen`? If `snprintf` went over the buffer length, you're already toast. The call to `strlen` will explode anyway. It seems safer to capture the return value of `snprintf` as you do elsewhere and work with that vs. `sizeof(timestamp_buf)`.

d2r replied:
Yes I agree with that. It seems better to check the result of `snprintf` here.

d2r commented on lib/ts/Diags.cc:
```diff
@@ -272,66 +267,31 @@ Diags::print_va(const char *debug_tag, DiagsLevel diags_level, const SourceLocat
-    *d++ = buffer[i];
+  format_writer.write(format_string, strlen(format_string));
+  format_writer.extend(2);
+  if (format_writer.data()[format_writer.extent() - 1] != '\n') {
```
Yes, `size()` is a better choice.

d2r commented:
Addressed review comments: 

```Diff
diff --git a/lib/ts/Diags.cc b/lib/ts/Diags.cc
index db8aa0fb4..abfa480b0 100644
--- a/lib/ts/Diags.cc
+++ b/lib/ts/Diags.cc
@@ -216,9 +216,8 @@ Diags::print_va(const char *debug_tag, DiagsLevel diags_level, const SourceLocat
 {
   ink_release_assert(diags_level < DiagsLevel_Count);
 
-  const size_t BUF_CAPACITY = 1024;
   using ts::LocalBufferWriter;
-  LocalBufferWriter<BUF_CAPACITY> format_writer;
+  LocalBufferWriter<1024> format_writer;
 
   // Save room for optional newline and terminating NUL bytes.
   format_writer.clip(2);
@@ -232,14 +231,16 @@ Diags::print_va(const char *debug_tag, DiagsLevel diags_level, const SourceLocat
     char timestamp_buf[48];
     char *buffer = ink_ctime_r(&cur_clock, timestamp_buf);
 
-    snprintf(&(timestamp_buf[19]), (sizeof(timestamp_buf) - 20), ".%03d", (int)(tp.tv_usec / 1000));
+    int num_bytes_written = snprintf(&(timestamp_buf[19]), (sizeof(timestamp_buf) - 20), ".%03d", (int)(tp.tv_usec / 1000));
 
-    format_writer.write('[');
-    format_writer.write(buffer + 4, std::min(strlen(buffer + 4), sizeof(timestamp_buf) - 4));
-    format_writer.write("] ", 2);
+    if (num_bytes_written > 0) {
+      format_writer.write('[');
+      format_writer.write(buffer + 4, num_bytes_written);
+      format_writer.write("] ", 2);
+    }
   }
 
-  size_t timestamp_end_offset = format_writer.extent();
+  size_t timestamp_end_offset = format_writer.size();
 
   ///////////////////////
   // add the thread id //
@@ -288,7 +289,7 @@ Diags::print_va(const char *debug_tag, DiagsLevel diags_level, const SourceLocat
 
   format_writer.write(format_string, strlen(format_string));
   format_writer.extend(2);
-  if (format_writer.data()[format_writer.extent() - 1] != '\n') {
+  if (format_writer.data()[format_writer.size() - 1] != '\n') {
     format_writer.write('\n');
   }
   format_writer.write('\0');
```

d2r commented:
Thanks for the reviews.

zwoop commented:
I think this should be reverted, and redone. At least two issues:


1) We should use the << operator that was added to the buffer writer, instead of repeated calls to write().

2) There's a bug in the timestamp, at a minimum this change needs to be done:

```
       format_writer.write('[');
-      format_writer.write(buffer + 4, num_bytes_written);
+      format_writer.write(buffer + 4, strlen(buffer) - 4);
```

However, a much better solution, IMO (and why I think this should be redone from scratch) is this:

```
diff --git a/lib/ts/Diags.cc b/lib/ts/Diags.cc
index abfa480b0..c54263ce7 100644
--- a/lib/ts/Diags.cc
+++ b/lib/ts/Diags.cc
@@ -34,6 +34,7 @@
 
  ****************************************************************************/
 
+#include <math.h>
 #include "ts/ink_platform.h"
 #include "ts/ink_memory.h"
 #include "ts/ink_defs.h"
@@ -231,13 +232,7 @@ Diags::print_va(const char *debug_tag, DiagsLevel diags_level, const SourceLocat
     char timestamp_buf[48];
     char *buffer = ink_ctime_r(&cur_clock, timestamp_buf);
 
-    int num_bytes_written = snprintf(&(timestamp_buf[19]), (sizeof(timestamp_buf) - 20), ".%03d", (int)(tp.tv_usec / 1000));
-
-    if (num_bytes_written > 0) {
-      format_writer.write('[');
-      format_writer.write(buffer + 4, num_bytes_written);
-      format_writer.write("] ", 2);
-    }
+    format_writer << '[' << buffer + 4 << floorf(tp.tv_usec * 1000) / 1000 << "] ";
   }
 
   size_t timestamp_end_offset = format_writer.size();
```

The above doesn't quite work, because of the buffer being \n l terminated I think, but you get the idea.

zwoop commented:
My point of this is that having replaced snprintf() with BufferWriter, only to create intermediate buffers with snprintf() seems completely nuts. Pick one.

zwoop commented:
Do we really need this for 7.1.x? This seems like a big change, and BufferWriter is different on 7.1.x vs master, isn't it?

d2r commented:
> Do we really need this for 7.1.x? This seems like a big change, and BufferWriter is different on 7.1.x vs master, isn't it?

`BufferWriter` was not present at all in 7.1.x, and so we did NOT want it to be back-ported there.

We thought it was best to request this go into a possible 7.2 release, since it was being discussed at the time.

If we are not doing a 7.2, then we can disregard the back-port altogether.
